checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt

```diff
@@ -0,0 +1,318 @@
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice and onUpdateLengthFilter event twice - when cvv is complete and valid and onUpdateCVV is called`() {
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV().onUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, true)
+        verify(cardListener, times(2)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice and onUpdateLengthFilter event twice - when cvv is invalid and onUpdateCVV is called`() {
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV().onUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+        verify(cardListener, times(2)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    @Test
+    fun `should not raise any events when cardValidator is not set and onUpdateCVV is called`() {
+        givenCvvIsInvalid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is partially valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice - when cvv is complete and valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is invalid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should not raise any events when cardValidator is not set and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onEndUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is on focus`() {
```

Comment:
Do you think we need to keep this test? This is essentially testing that if there's not listener then the listener is not called, right?

Reply:
Come to think of it, no we don't. This test was added to ensure we do not use the cardListener as a null, avoiding the NPE. However in Kotlin, it fails to compile anyway if you try to do this. It is more  of a test case if this was in Java so I can remove this test.